### PR TITLE
Clean up SyncJobStatus and add FhirSynchronizer test

### LINKED ISSUE #2183: Add tests to sync module classes
**Describe the Issue**

**Would you like to work on the issue?**

### PR DESCRIPTION
**Pull request by omarismail94**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #2183 

**Description**

* This PR introduces unit tests for FhirSynchronizer, ensuring that the synchronisation process functions as expected under various conditions.
* Adding tests led to realising that we can make the `SyncJobStatus.Started` and `SyncJobStatus.Finished` objects as opposed to being instantiated each time
* Create a `FhirSynchronizer.syncState` property that a caller can listen to without having to tell `FhirSynchronizer`

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | **Code health** | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: demo/src/main/java/com/google/android/fhir/demo/MainActivityViewModel.kt [hunk 1]
```diff
@@ -48,34 +49,39 @@ class MainActivityViewModel(application: Application) : AndroidViewModel(applica
...
     }
   }
 
   fun triggerOneTimeSync() {
-    viewModelScope.launch {
-      Sync.oneTimeSync<DemoFhirSyncWorker>(getApplication())
-        .shareIn(this, SharingStarted.Eagerly, 10)
-        .collect { _pollState.emit(it) }
-    }
+    syncJob?.cancel()
```
Comment (anchita-g) [inline demo/src/main/java/com/google/android/fhir/demo/MainActivityViewModel.kt:70]: Why do we want to cancel the existing running Job? Could we not enqueue the new job to be run after the current one finishes?

Comment (omarismail94) [reply demo/src/main/java/com/google/android/fhir/demo/MainActivityViewModel.kt:70]: I found that the issue with multiple emissions was due to the creation of a new collector each time the `triggerOneTimeSync` function is invoked. Each of these collectors independently subscribes to the `SharedFlow` and begins emitting values. As a result, when the function is called N times, we end up with N active collectors, each emitting their own sequence of values. This leads to the observed behaviour where we see multiple emissions for a single event in the logs

By canceling the existing job before setting up a new collector, we ensure that at any given time, there's only one active collector and prioritise the latest sync request

Comment (anchita-g) [reply demo/src/main/java/com/google/android/fhir/demo/MainActivityViewModel.kt:70]: Cancelling the existing job would imply that the cancellation of the job does not leave resources in an incoherent state. 
Would an enqueued job still initiate a collector even though it is not running?

Comment (omarismail94) [reply demo/src/main/java/com/google/android/fhir/demo/MainActivityViewModel.kt]: I removed this code from this PR and opened up an issue (https://github.com/google/android-fhir/issues/2191) refering to this thread. Will come back to this once we complete the re-design of the SyncJobStatus API

**Review by aditya-07 — COMMENTED** [review]
Just this one comment, otherwise the change looks fine.

**Review by jingtang10 — APPROVED** [review]
the changes itself in this PR are net positive - so i'm happy for this to be merged :) and thanks so much omar for improving our code health!

but please create a new issue if you don't want to aditya's comment in this pr.

File: engine/src/test/java/com/google/android/fhir/sync/FhirSynchronizerTest.kt [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        TestFhirEngineImpl,
+        uploader,
+        downloader,
+        conflictResolver,
+      )
+  }
+
+  @Test
+  fun `subscribe should throw exception on second subscription`() {
+    val flow1 = MutableSharedFlow<SyncJobStatus>()
```
[the remark this answers is not in the document]
Comment (omarismail94) [reply engine/src/test/java/com/google/android/fhir/sync/FhirSynchronizerTest.kt]: added!

**Pull request merged**